htlcswitch/link_test.go

```diff
@@ -277,125 +275,93 @@ func TestChannelLinkBidirectionalOneHopPayments(t *testing.T) {
-		channels.bobToCarol, channels.carolToBob, testStartingHeight)
+	n := newTwoHopNetwork(
+		t, channels.aliceToBob, channels.bobToAlice, testStartingHeight,
+	)
 	if err := n.start(); err != nil {
 		t.Fatal(err)
 	}
 	defer n.stop()
-	bobBandwidthBefore := n.firstBobChannelLink.Bandwidth()
-	aliceBandwidthBefore := n.aliceChannelLink.Bandwidth()
-
-	debug := false
-	if debug {
-		// Log message that alice receives.
-		n.aliceServer.intersect(createLogFunc("alice",
-			n.aliceChannelLink.ChanID()))
-
-		// Log message that bob receives.
-		n.bobServer.intersect(createLogFunc("bob",
-			n.firstBobChannelLink.ChanID()))
-	}
 
+	// Fill up the commitment from Alice's side with 20 sat payments.
+	count := (input.MaxHTLCNumber / 2)
 	amt := lnwire.NewMSatFromSatoshis(20000)
 
 	htlcAmt, totalTimelock, hopsForwards := generateHops(amt,
-		testStartingHeight, n.firstBobChannelLink)
-	_, _, hopsBackwards := generateHops(amt,
-		testStartingHeight, n.aliceChannelLink)
+		testStartingHeight, n.bobChannelLink)
 
-	type result struct {
-		err    error
-		start  time.Time
-		number int
-		sender string
-	}
+	firstHop := n.aliceChannelLink.ShortChanID()
+
+	// Create channels to buffer the preimage and error channels used in
+	// making the preliminary payments.
+	alicePreimageChan := make(chan lntypes.Preimage, count)
+	aliceErrChan := make(chan chan error, count)
 
-	// Send max available payment number in both sides, thereby testing
-	// the property of channel link to cope with overflowing.
-	count := 2 * input.MaxHTLCNumber
-	resultChan := make(chan *result, count)
-	for i := 0; i < count/2; i++ {
+	var wg sync.WaitGroup
+	for i := 0; i < count; i++ {
+		var preimage lntypes.Preimage
+		_, err := rand.Read(preimage[:])
+		if err != nil {
+			t.Fatalf("unable to generate preimage: %v", err)
+		}
+
+		wg.Add(1)
 		go func(i int) {
-			r := &result{
-				start:  time.Now(),
-				number: i,
-				sender: "alice",
-			}
+			defer wg.Done()
 
-			firstHop := n.firstBobChannelLink.ShortChanID()
-			_, r.err = makePayment(
+			err := n.makeHoldPayment(
 				n.aliceServer, n.bobServer, firstHop,
 				hopsForwards, amt, htlcAmt, totalTimelock,
-			).Wait(5 * time.Minute)
-			resultChan <- r
+				preimage,
+			)
+			alicePreimageChan <- preimage
+			aliceErrChan <- err
 		}(i)
 	}
 
-	for i := 0; i < count/2; i++ {
-		go func(i int) {
-			r := &result{
-				start:  time.Now(),
-				number: i,
-				sender: "bob",
-			}
+	// Wait for Alice to finish filling her commitment.
+	wg.Wait()
+	close(alicePreimageChan)
+	close(aliceErrChan)
 
-			firstHop := n.aliceChannelLink.ShortChanID()
-			_, r.err = makePayment(
-				n.bobServer, n.aliceServer, firstHop,
-				hopsBackwards, amt, htlcAmt, totalTimelock,
-			).Wait(5 * time.Minute)
-			resultChan <- r
-		}(i)
+	// Now make an additional payment from Alice to Bob, this should be
+	// canceled because the commitment in this direction is full.
+	err = <-makePayment(
```

Comment:
Hmm, maybe a (more) sync version `makePayment` would be better. The reliance on the err being sent immediately isn't instantly clear

Reply:
what do you mean by sent immediately?